File: src/main/java/frc/robot/RobotContainer.java
package frc.robot;

import com.pathplanner.lib.auto.AutoBuilder;
import com.pathplanner.lib.commands.PathPlannerAuto;
import com.pathplanner.lib.util.FlippingUtil;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.wpilibj.Filesystem;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;
import edu.wpi.first.wpilibj2.command.button.CommandXboxController;
import frc.robot.commands.DriveCommands;
import frc.robot.subsystems.drive.Drive;
import frc.robot.subsystems.drive.DriveConstants;
import frc.robot.subsystems.drive.gyro.GyroIO;
import frc.robot.subsystems.drive.gyro.GyroIOPigeon2;
import frc.robot.subsystems.drive.gyro.GyroIOSim;
import frc.robot.subsystems.drive.module.ModuleIO;
import frc.robot.subsystems.drive.module.ModuleIOSim;
import frc.robot.subsystems.drive.module.ModuleIOTalonFX;
import java.io.File;
import java.util.Objects;
import org.ironmaple.simulation.SimulatedArena;
import org.ironmaple.simulation.drivesims.SwerveDriveSimulation;
import org.littletonrobotics.junction.networktables.LoggedDashboardChooser;

public class RobotContainer {
  // Controllers (bind both to the same port in sim for ease of control)
  private final CommandXboxController driver = new CommandXboxController(0);

  // Subsystems
  private final Drive drivetrain;

  // Elastic auto chooser
  private LoggedDashboardChooser<Command> autoChooser;

  // MapleSim configuration (start on the field)
  public static SwerveDriveSimulation swerveDriveSimulation =
      new SwerveDriveSimulation(
          DriveConstants.driveTrainSimulationConfig, new Pose2d(3, 3, Rotation2d.kZero));

  public RobotContainer() {
    switch (Constants.currentMode) {
      case REAL:
        // Real robot, instantiate hardware IO implementations
        drivetrain =
            new Drive(
                new GyroIOPigeon2(),
                new ModuleIOTalonFX(DriveConstants.moduleConfigs[0]),
                new ModuleIOTalonFX(DriveConstants.moduleConfigs[1]),
                new ModuleIOTalonFX(DriveConstants.moduleConfigs[2]),
                new ModuleIOTalonFX(DriveConstants.moduleConfigs[3]),
                (robotPose) -> {});
        break;

      case SIM:
        // Sim robot, instantiate physics sim IO implementations
        drivetrain =
            new Drive(
                new GyroIOSim(swerveDriveSimulation.getGyroSimulation()),
                new ModuleIOSim(swerveDriveSimulation.getModules()[0]),
                new ModuleIOSim(swerveDriveSimulation.getModules()[1]),
                new ModuleIOSim(swerveDriveSimulation.getModules()[2]),
                new ModuleIOSim(swerveDriveSimulation.getModules()[3]),
                (robotPose) -> swerveDriveSimulation.setSimulationWorldPose(robotPose));
        drivetrain.setPose(swerveDriveSimulation.getSimulatedDriveTrainPose());
        break;

      default:
        // Replayed robot, disable IO implementations
        drivetrain =
            new Drive(
                new GyroIO() {},
                new ModuleIO() {},
                new ModuleIO() {},
                new ModuleIO() {},
                new ModuleIO() {},
                (robotPose) -> {});
        break;
    }

    // Add drivetrain to MapleSim if in simulation
    if (RobotBase.isSimulation())
      SimulatedArena.getInstance().addDriveTrainSimulation(swerveDriveSimulation);

    setupControllerBindings();
    setupPathPlanner();
  }

  private void setupControllerBindings() {

    // DRIVE CONTROLLER
    drivetrain.setDefaultCommand(
        DriveCommands.fieldRelative(
            drivetrain,
            () -> -driver.getLeftY(),
            () -> -driver.getLeftX(),
            () -> -driver.getRightX()));

    // Left bumper: Robot relative
    driver
        .leftBumper()
        .whileTrue(
            DriveCommands.robotRelative(
                drivetrain,
                () -> -driver.getLeftY(),
                () -> -driver.getLeftX(),
                () -> -driver.getRightX()));

    // A: X-brake
    driver.a().whileTrue(DriveCommands.brakeWithX(drivetrain));

    // Left trigger: Align to closest left branch
    driver
        .leftTrigger()
        .whileTrue(
            DriveCommands.alignToClosestReefBranch(
                drivetrain,
                DriveCommands.BranchSide.LEFT,
                () -> -driver.getLeftY(),
                () -> -driver.getLeftX()));

    // Right trigger: Align to closest right branch
    driver
        .rightTrigger()
        .whileTrue(
            DriveCommands.alignToClosestReefBranch(
                drivetrain,
                DriveCommands.BranchSide.RIGHT,
                () -> -driver.getLeftY(),
                () -> -driver.getLeftX()));

    // Reset gyro to 0 when start button is pressed.
    driver
        .start()
        .onTrue(
            Commands.runOnce(
                    () ->
                        drivetrain.setPose(
                            new Pose2d(
                                drivetrain.getPose().getTranslation(),
                                Rotation2d.kZero.plus(
                                    AutoBuilder.shouldFlip() ? Rotation2d.kPi : Rotation2d.kZero))),
                    drivetrain)
                .ignoringDisable(true));

    // Feedforward characterization
    driver.x().whileTrue(DriveCommands.feedforwardCharacterization(drivetrain));

    // Wheel radius characterization
    driver.b().whileTrue(DriveCommands.wheelRadiusCharacterization(drivetrain));
  }

  private void setupPathPlanner() {
    // Auto chooser
    autoChooser = new LoggedDashboardChooser<>("Selected Auto");
    File autosFolder = new File(Filesystem.getDeployDirectory() + "/pathplanner/autos");
    for (File file : Objects.requireNonNull(autosFolder.listFiles())) {
      if (!file.getName().endsWith(".auto") || file.getName().equals(".DS_Store")) continue;
      String name = file.getName().replace(".auto", "");
      autoChooser.addOption(name, new PathPlannerAuto(name, false));
      autoChooser.addOption(name + " (Mirrored)", new PathPlannerAuto(name, true));
    }
    autoChooser.addDefaultOption("None", Commands.none());
    autoChooser.addOption(
        "Basic Leave",
        drivetrain
            .run(() -> DriveCommands.robotRelative(drivetrain, () -> 0.25, () -> 0, () -> 0))
            .withTimeout(1.0));
  }

  public void readyRobotForMatch() {
    if (autoChooser.get() instanceof PathPlannerAuto auto) {
      if (AutoBuilder.shouldFlip())
        drivetrain.setPose(FlippingUtil.flipFieldPose(auto.getStartingPose()));
      else drivetrain.setPose(auto.getStartingPose());
    }
  }

  public Command getAutonomousCommand() {
    return autoChooser.get();
  }
}

File: python/robot_container.py
import math
import os.path

import commands2.button
from commands2 import cmd, Command
from commands2.button import CommandXboxController
from pathplannerlib.auto import AutoBuilder, PathPlannerAuto
from pathplannerlib.util import FlippingUtil
from pykit.networktables.loggeddashboardchooser import LoggedDashboardChooser
from wpilib import getDeployDirectory
from wpimath.geometry import Pose2d, Rotation2d

import commands
from commands import fieldRelative
from constants import Constants
from subsystems.drive import Drive
from subsystems.drive.gyro import GyroIOPigeon2, GyroIOSim
from subsystems.drive.module import ModuleIOTalonFX, ModuleIOSim
from subsystems.toast import moduleConfigs
from subsystems.vision import Vision
from subsystems.vision.io import VisionIOLimelight


class RobotContainer:
    def __init__(self) -> None:
        self._driver = CommandXboxController(0)

        match Constants.currentMode:
            case Constants.Mode.REAL:
                # Real robot, instantiate hardware IO implementations
                self._drivetrain = Drive(
                    GyroIOPigeon2(),
                    ModuleIOTalonFX(moduleConfigs[0]),
                    ModuleIOTalonFX(moduleConfigs[1]),
                    ModuleIOTalonFX(moduleConfigs[2]),
                    ModuleIOTalonFX(moduleConfigs[3]),
                    lambda _: None
                )
                self._vision = Vision(
                    self._drivetrain.addVisionMeasurement,
                    VisionIOLimelight("limelight", self._drivetrain.getRotation)
                )
            case Constants.Mode.SIM:
                # Sim robot, instantiate physics sim IO implementations (if available)
                self._drivetrain = Drive(
                    GyroIOSim(),
                    ModuleIOSim(moduleConfigs[0]),
                    ModuleIOSim(moduleConfigs[1]),
                    ModuleIOSim(moduleConfigs[2]),
                    ModuleIOSim(moduleConfigs[3]),
                    lambda _: None
                )
                self._vision = Vision(
                    self._drivetrain.addVisionMeasurement,
                    VisionIOLimelight("limelight", self._drivetrain.getRotation)
                )

        # Auto chooser
        self.autoChooser: LoggedDashboardChooser[Command] = LoggedDashboardChooser("Selected Auto")

        auto_files = os.listdir(os.path.join(getDeployDirectory(), "pathplanner", "autos"))
        for file in auto_files:
            file = file.removesuffix(".auto")
            self.autoChooser.addOption(file, PathPlannerAuto(file, False))
            self.autoChooser.addOption(f"{file} (Mirrored)", PathPlannerAuto(file, True))
        self.autoChooser.setDefaultOption("None", cmd.none())
        self.autoChooser.addOption("Basic Leave",
        self._drivetrain.run(lambda: commands.robotRelative(self._drivetrain, lambda: 0.25, lambda: 0, lambda: 0)).withTimeout(1.0))

        self.setupControllerBindings()

    def readyRobotForMatch(self) -> None:
        auto = self.autoChooser.getSelected()
        if isinstance(auto, PathPlannerAuto):
            if AutoBuilder.shouldFlip():
                self._drivetrain.setPose(FlippingUtil.flipFieldPose(auto._startingPose))
            else:
                self._drivetrain.setPose(auto._startingPose)

    def setupControllerBindings(self) -> None:
        # DRIVE CONTROLLER
        self._drivetrain.setDefaultCommand(fieldRelative(
            self._drivetrain,
            lambda: -self._driver.getLeftY(),
            lambda: -self._driver.getLeftX(),
            lambda: -self._driver.getRightX()
        ))

        # Left bumper: Robot relative
        self._driver.leftBumper().whileTrue(
            commands.robotRelative(
                self._drivetrain,
                lambda: -self._driver.getLeftY(),
                lambda: -self._driver.getLeftX(),
                lambda: -self._driver.getRightX()
            )
        )

        # A: X-brake
        self._driver.a().whileTrue(commands.brakeWithX(self._drivetrain))

        # Left trigger: Align to closest left branch
        self._driver.leftTrigger().whileTrue(
            commands.alignToClosestBranch(
                self._drivetrain,
                commands.BranchSide.LEFT,
                lambda: -self._driver.getLeftY(),
                lambda: -self._driver.getLeftX()
            )
        )

        # Right trigger: Align to closest right branch
        self._driver.rightTrigger().whileTrue(
            commands.alignToClosestBranch(
                self._drivetrain,
                commands.BranchSide.RIGHT,
                lambda: -self._driver.getLeftY(),
                lambda: -self._driver.getLeftX()
            )
        )

        # Reset gyro to 0 when start button is pressed.
        self._driver.start().onTrue(cmd.runOnce(
            lambda: self._drivetrain.setPose(
                Pose2d(self._drivetrain.getPose().translation(), Rotation2d() + Rotation2d(math.pi) if AutoBuilder.shouldFlip() else Rotation2d())
            ), self._drivetrain
        ).ignoringDisable(True))

        # Feedforward characterization
        self._driver.x().whileTrue(commands.feedforwardCharacterization(self._drivetrain))

        # Wheel radius characterization
        self._driver.b().whileTrue(commands.wheelRadiusCharacterization(self._drivetrain))

    def get_autonomous_command(self) -> commands2.Command:
        return self.autoChooser.getSelected()
